buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalBwcGitPlugin.java

```diff
@@ -0,0 +1,202 @@
+
+import static java.util.Arrays.asList;
+
+public class InternalBwcGitPlugin implements Plugin<Project> {
+
+    private BwcGitExtension gitExtension;
+    private Project project;
+
+    @Override
+    public void apply(Project project) {
+        this.project = project;
+        this.gitExtension = project.getExtensions().create("bwcGitConfig", BwcGitExtension.class);
+        ProviderFactory providers = project.getProviders();
+        Provider<String> remote = project.getProviders().systemProperty("bwc.remote").forUseAtConfigurationTime().orElse("elastic");
+
+        TaskContainer tasks = project.getTasks();
+        TaskProvider<LoggedExec> createCloneTaskProvider = tasks.register("createClone", LoggedExec.class, createClone -> {
+            createClone.onlyIf(task -> this.gitExtension.getCheckoutDir().get().exists() == false);
+            createClone.setCommandLine(asList("git", "clone", project.getRootDir(), gitExtension.getCheckoutDir().get()));
+        });
+
+        TaskProvider<LoggedExec> findRemoteTaskProvider = tasks.register("findRemote", LoggedExec.class, findRemote -> {
+            findRemote.dependsOn(createCloneTaskProvider);
+            // TODO Gradle should provide property based configuration here
+            findRemote.setWorkingDir(gitExtension.getCheckoutDir().get());
+
+            findRemote.setCommandLine(asList("git", "remote", "-v"));
+            ByteArrayOutputStream output = new ByteArrayOutputStream();
+            findRemote.setStandardOutput(output);
+            findRemote.doLast(t -> {
+                ExtraPropertiesExtension extraProperties = project.getExtensions().getExtraProperties();
+                extraProperties.set("remoteExists", isRemoteAvailable(remote, output));
+            });
+        });
+
+        TaskProvider<LoggedExec> addRemoteTaskProvider = tasks.register("addRemote", LoggedExec.class, addRemote -> {
+            addRemote.dependsOn(findRemoteTaskProvider);
+            addRemote.onlyIf(task -> ((boolean) project.getExtensions().getExtraProperties().get("remoteExists")) == false);
+            addRemote.setWorkingDir(gitExtension.getCheckoutDir().get());
+            String remoteRepo = remote.get();
+            // for testing only we can override the base remote url
+            String remoteRepoUrl = project.getProviders()
+                .systemProperty("testRemoteRepo")
+                .forUseAtConfigurationTime()
+                .getOrElse("https://github.com/" + remoteRepo + "/elasticsearch.git");
+            addRemote.setCommandLine(asList("git", "remote", "add", remoteRepo, remoteRepoUrl));
+        });
+
+        TaskProvider<LoggedExec> fetchLatestTaskProvider = tasks.register("fetchLatest", LoggedExec.class, fetchLatest -> {
+            var gitFetchLatest = project.getProviders()
+                .systemProperty("tests.bwc.git_fetch_latest")
+                .forUseAtConfigurationTime()
+                .orElse("true")
+                .map(fetchProp -> {
```

Comment:
Ideally we should move away from just using System.getProperty and therefore Util.getBooleanProperty as is as this is discouraged by gradle in order to support `--configuration-cache`. IMO we should actually deprecate `Util.getBooleanProperty`

Reply:
Understood. Yeah, we do this pattern a lot so maybe we should later add some helper for returning a provider.

Reply:
Why not turn `Util.getBooleanProperty()` into that helper?

Reply:
I don't believe everywhere we use that is setup to consume a `Provider` as-is. We'd have to do some refactoring there and it's significant to probably address in a separate PR.

Reply:
Sorry, I was not suggesting changing in this PR, only changing the implementation of that method in a followup, rather than continuing to add any more uses of this pattern.